### PR TITLE
Fix CGMES import with InputStream

### PR DESCRIPTION
**Pull request by colinepiloquet**
Signed-off-by: Coline PILOQUET <coline.piloquet@rte-france.com>

**Please check if the PR fulfills these requirements** *(please use `'[x]'` to check the checkboxes, or submit the PR and then click the checkboxes)*
- [x] The commit message follows our guidelines
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**Does this PR already have an issue describing the problem ?** *If so, link to this issue using `'#XXX'` and skip the rest*
No

**What kind of change does this PR introduce?** *(Bug fix, feature, docs update, ...)*
Bug fix

**What is the current behavior?** *(You can also link to an open issue here)*
The import of CGMES files fails when using loadNetwork(String filename, InputStream data)

### CODE REVIEW

File: commons/src/main/java/com/powsybl/commons/datasource/ZipReadOnlyMemDataSource.java [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+
+    @Override
+    public Set<String> listNames(String regex) throws IOException {
+        Pattern p = Pattern.compile(regex);
+        return data.keySet().stream()
+                .filter(name -> p.matcher(name).matches())
+                .collect(Collectors.toSet());
+    }
+
+    private static final class ZipEntryInputStream extends ForwardingInputStream<InputStream> {
```
Comment (flo-dup) [inline commons/src/main/java/com/powsybl/commons/datasource/ZipReadOnlyMemDataSource.java:100]: you can delete this as you're not using it

File: commons/src/main/java/com/powsybl/commons/datasource/ZipReadOnlyMemDataSource.java [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+import java.util.zip.ZipEntry;
+import java.util.zip.ZipInputStream;
+
+public class ZipReadOnlyMemDataSource implements ReadOnlyDataSource {
```
Comment (flo-dup) [inline commons/src/main/java/com/powsybl/commons/datasource/ZipReadOnlyMemDataSource.java:20]: Should we add this new class which is almost the same as ReadOnlyMemDataSource? The only difference is `putData`. I think we could have reused ReadOnlyMemDataSource in fact, and put the code you wrote in `putData` in `DataSourceUtil`, like the already existing one for file. Or we could have extended ReadOnlyMemDataSource maybe? what do you think?

**Review by flo-dup — CHANGES_REQUESTED** [review]

File: commons/src/main/java/com/powsybl/commons/datasource/ZipReadOnlyMemDataSource.java (hunk 2 above)
Comment (colinepiloquet) [reply commons/src/main/java/com/powsybl/commons/datasource/ZipReadOnlyMemDataSource.java:18]: I actually agree, it should be possible to use only ReadOnlyMemDataSource. I will try to add the code in DataSourceUtil.

File: iidm/iidm-converter-api/src/main/java/com/powsybl/iidm/import_/Importers.java [hunk 3]
```diff
@@ -549,8 +550,16 @@ public static Network loadNetwork(String file) {
      * @return                   The loaded network
      */
     public static Network loadNetwork(String filename, InputStream data, ComputationManager computationManager, ImportConfig config, Properties parameters, ImportersLoader loader, Reporter reporter) {
-        ReadOnlyMemDataSource dataSource = new ReadOnlyMemDataSource(DataSourceUtil.getBaseName(filename));
-        dataSource.putData(filename, data);
+        ReadOnlyDataSource dataSource;
+        if (filename.endsWith(".zip")) {
```
[the remark this answers is not in the document]
Comment (colinepiloquet) [reply iidm/iidm-converter-api/src/main/java/com/powsybl/iidm/import_/Importers.java:554]: I didn't think about the other kinds of compressed files! But I will add a condition

**Review by geofjamg — CHANGES_REQUESTED** [review]
I really doubt this is the right way to solve the issue. The initial need if I understand well is to read a zip file from an `InputStream`, but not to buffer all the zip data. CGMES files which are the main purpose of this feature could be like 1Go deflatted. Is there another way to solve this issue, without buffering everything?

**Comment by flo-dup** [comment]
I don't really see how we can improve that. Indeed, we start with an InputStream and we need to go from one zip entry to the other, not only onward but also backward. Therefore we need to copy the stream, hence buffering it in a way or another.

Ah, but maybe the solution would be writing a temporary zip file and use `ZipFileDataSource`?

**Comment by geofjamg** [comment]
> I don't really see how we can improve that. Indeed, we start with an InputStream and we need to go from one zip entry to the other, not only onward but also backward. Therefore we need to copy the stream, hence buffering it in a way or another.
> 
> Ah, but maybe the solution would be writing a temporary zip file and use `ZipFileDataSource`?

But we can reset position of an inputstream?

**Comment by flo-dup** [comment]
Yeah it's probably the best way to do, not supporting a zipped inputStream and throwing a clear exception.

**Comment by flo-dup** [comment]
Well, we can reset an InputStream only if there's some buffering done behind. I don't know exactly, maybe it's better than inflating all the files.

If I remember correctly there are quite many InputStreams created when reading a CGMES file, for each file more than once. Like it's done once for all for checking the namespaces for instance.

**Pull request closed without merging**